File: data-structures/ocaml/bin/btree.ml
type ('k, 'v) pair = ('k * 'v)
type ('k, 'v) node = {
  mutable length : int;
  mutable items : ('k, 'v) pair list;
  mutable children : ('k, 'v) node list option;
}

let make_node is_leaf =
  if is_leaf then
    {
      length = 0;
      items = [];
      children = None;
    }
  else
    {
      length = 0;
      items = [];
      children = Some [];
    }

let set_node_length node length =
  match node with
  | None -> ()
  | Some node ->
    node.length <- length

let set_node_items node items =
  match node with
  | None -> ()
  | Some node ->
    node.items <- items

let node_is_leaf node =
  match node with
  | None -> true
  | Some node ->
    match node.children with
    | None -> true
    | Some _ -> false

let rec list_set_item lst index item =
  match lst with
  | [] -> []
  | hd :: tl ->
    if index = 0 then
      item :: tl
    else
      hd :: list_set_item tl (index - 1) item

let search node key =
  let rec binary_search low high =
    if low < high then
      let h = (low + high) / 2 in
      let item = List.nth node.items h in
      if key >= fst item then
        binary_search (h + 1) high
      else
        binary_search low h
    else
      low
  in
  let low = binary_search 0 (List.length node.items) in
  if low > 0 then
    let item = List.nth node.items (low - 1) in
    if fst item >= key then
      (low - 1, true)
    else
      (low, false)
  else
    (low, false)

let degree_to_min_max degree =
  let degree =
    if degree < 0 then
      32
    else if degree == 1 then
      2
    else degree
  in
  let max = 2 * degree - 1 in
  let min = max / 2 in
  (min, max)

module Btree : sig
  type ('k, 'v) t = {
    mutable root : ('k, 'v) node option;
    mutable length : int;
    mutable min : int;
    mutable max : int;
  }

  val create : int -> ('k, 'v) t

  val insert : ('k, 'v) t -> ('k * 'v) -> ('k, 'v) t

end = struct
  type ('k, 'v) t = {
    mutable root : ('k, 'v) node option;
    mutable length : int;
    mutable min : int;
    mutable max : int;
  }

  let create degree =
    let (min, max) = degree_to_min_max degree in
    {
      root = None;
      length = 0;
      min;
      max;
    }

  let split_node btree node =
    failwith "TODO"

  let rec insert_node btree parent item =
    let key = fst item in
    let (i, found) = search parent key in
    if found then
      let item = List.nth parent.items i in
      let prev = snd item in
      let list_items = list_set_item parent.items i item in
      set_node_items (Some parent) list_items;
      (Some prev, true, false)
    else
      if node_is_leaf (Some parent) then
        let item_length = List.length parent.items in
        if item_length = btree.max then
          (None, false, true)
        else
          (* TODO: It should be improve to assign only once, instead of two mutations. *)
          let new_items = item :: parent.items in
          set_node_items (Some parent) new_items;
          let list_items = list_set_item parent.items i item in
          set_node_items (Some parent) list_items;
          parent.length <- parent.length + 1;
          (None, false, false)
      else
        let child =
          match parent.children with
          (* TODO: Validate a scenario where the parent is a leaf node and the children is not. *)
          | None -> failwith "TODO - missing children on parent"
          | Some children -> List.nth children i
        in
        let (prev, replaced, split) = insert_node btree child item in
        if split then
          let items_length = List.length parent.items in
          if items_length = btree.max then
            (None, false, true)
          else (
            let (right, median) = split_node child in
            parent.children <- None;
            failwith "TODO - split_node"
          )
        else
          if not replaced then
            let () = set_node_length (Some parent) (parent.length + 1) in
            (prev, replaced, false)
          else
            (prev, replaced, false)

  let rec insert btree (key, value) =
    let item = (key, value) in
    match btree.root with
    | None ->
      btree.root <- Some (make_node true);
      set_node_length btree.root 1;
      set_node_items btree.root [item];
      btree.length <- 1;
      btree
    | Some node ->
      let (prev, found, split) = insert_node btree node item in
      if split then
        let left = btree.root in
        let (right, median) = split_node left in
        failwith "TODO"
      else
        failwith "TODO"
end

let () =
  let _btree = Btree.create 32 in
  let _ = Btree.insert _btree (1, "one") in
  let _ = Btree.insert _btree (1, "two") in
  ()

File: data-structures/ts/src/btree.ts
type BTreePair<K, V> = {
  key: K;
  value: V;
}

type BTreeNode<K, V> = {
  length: number;
  items: Array<BTreePair<K, V> | null>;
  children: Array<BTreeNode<K, V>>;
};

type BTree<K, V> = {
  root: BTreeNode<K, V> | null;
  length: number;
  min: number;
  max: number;
}

export function btreeCreate<K, V>(degree?: number): BTree<K, V> {
  const [min, max] = degreeToMinMax(degree || 0);
  return {
    root: null,
    length: 0,
    min,
    max,
  }
}

export function btreeInsert<K, V>(btree: BTree<K, V>, key: K, value: V) {
  const item: BTreePair<K, V> = { key, value };
  if (btree.root === null) {
    btree.root = makeNode(true);
    btree.root.length = 1;
    btree.root.items = [item];
    btree.length = 1;
    return [null, false];
  }
  const [prev, replaced, split] = insertNode(btree, btree.root, item);
  if (split) {
    const left = btree.root;
    const [right, median] = splitNode(btree, left);
    btree.root = makeNode(false);
    btree.root.children = [left, right];
    btree.root.items = [median];
    updateCount(btree.root);
    return btreeInsert(btree, key, value);
  }
  if (replaced) {
    return [prev, true];
  }
  btree.length++;
  return [null, false];
}

const insertNode = <K, V>(btree: BTree<K, V>, parent: BTreeNode<K, V>, item: BTreePair<K, V>): [BTreePair<K, V> | null, boolean, boolean] => {
  const [i, found] = binarySearch(parent, item.key);
  if (found) {
    const prev = parent.items[i];
    parent.items[i] = item;
    return [prev, true, false];
  }
  if (nodeIsLeaf(parent)) {
    const itemLength = parent.items.length;
    if (itemLength == btree.max) {
      return [null, false, true];
    }
    parent.items.splice(i, 0, item);
    parent.length++;
    return [null, false, false];
  }
  const [prev, replaced, split] = insertNode(btree, parent.children[i], item);
  if (split) {
    const itemsLength = parent.items.length;
    if (itemsLength == btree.max) {
      return [null, false, true];
    }
    const child = parent.children![i];
    const [right, median] = splitNode(btree, child);
    parent.children.splice(i + 1, 0, right);
    parent.items.splice(i, 0, median);
    return insertNode(btree, parent, item);
  }
  if (replaced) {
    parent.length++;
  }
  return [prev, replaced, false];
}

const splitNode = <K, V>(btree: BTree<K, V>, node: BTreeNode<K, V>): [BTreeNode<K, V>, BTreePair<K, V>] => {
  const i = Math.floor(btree.max / 2);
  const median = node.items[i];

  const right = makeNode<K, V>(nodeIsLeaf(node));
  right.items = node.items.splice(i + 1);

  if (!nodeIsLeaf(node)) {
    right.children = node.children.splice(i + 1);
  }

  updateCount(right);

  node.items = node.items.splice(0, i);
  if (!nodeIsLeaf(node)) {
    node.children = node.children.splice(0, i + 1);
  }
  updateCount(node);

  return [right, median!];
}

const degreeToMinMax = (degree: number): [number, number] => {
  const _degree = degree <= 0 ? 32 : degree == 1 ? 2 : degree;
  const max = 2 * _degree - 1;
  const min = Math.floor(max / 2);
  return [min, max];
}

const makeNode = <K, V>(isLeaf: boolean): BTreeNode<K, V> => {
  return {
    length: 0,
    items: [],
    children: [],
  }
}

const nodeIsLeaf = <K, V>(node: BTreeNode<K, V>): boolean =>
  node.children.length === 0;

const updateCount = <K, V>(node: BTreeNode<K, V>) => {
  node.length = node.items.length;
  if (!nodeIsLeaf(node)) {
    if (!node.children) {
      return;
    }

    for (const child of node.children) {
      updateCount(child);
    }
  }
}

const binarySearch = <K, V>(node: BTreeNode<K, V>, key: K): [number, boolean] => {
  let low = 0;
  let high = node.items.length;
  while (low < high) {
    const h = Math.floor((low + high) / 2);
    const item = node.items[h];
    if (!(key < item!.key)) {
      low = h + 1;
    } else {
      high = h;
    }
  }
  const item = node.items[low - 1];
  if (low > 0 && !(item!.key < key)) {
    return [low - 1, true];
  }
  return [low, false];
}
